core/src/main/kotlin/gateway/handler/GatewayEventInterceptor.kt

```diff
@@ -1,51 +1,9 @@
 package dev.kord.core.gateway.handler
 
-import dev.kord.cache.api.DataCache
-import dev.kord.core.Kord
-import dev.kord.core.gateway.MasterGateway
-import dev.kord.gateway.Event
-import io.ktor.util.*
-import kotlinx.coroutines.channels.Channel
-import kotlinx.coroutines.flow.MutableSharedFlow
-import kotlinx.coroutines.flow.buffer
-import kotlinx.coroutines.flow.launchIn
-import kotlinx.coroutines.flow.onEach
-import mu.KotlinLogging
-import dev.kord.core.event.Event as CoreEvent
+import kotlinx.coroutines.Job
 
-private val logger = KotlinLogging.logger { }
+interface GatewayEventInterceptor {
 
-@Suppress("EXPERIMENTAL_API_USAGE")
-class GatewayEventInterceptor(
-    private val kord: Kord,
-    private val gateway: MasterGateway,
-    cache: DataCache,
-    coreFlow: MutableSharedFlow<CoreEvent>,
-) {
+    suspend fun start(): Job
 
-    private val listeners = listOf(
-        MessageEventHandler(kord, gateway, cache, coreFlow),
-        ChannelEventHandler(kord, gateway, cache, coreFlow),
-        ThreadEventHandler(kord, gateway, cache, coreFlow),
-        GuildEventHandler(kord, gateway, cache, coreFlow),
-        LifeCycleEventHandler(kord, gateway, cache, coreFlow),
-        UserEventHandler(kord, gateway, cache, coreFlow),
-        VoiceEventHandler(kord, gateway, cache, coreFlow),
-        WebhookEventHandler(kord, gateway, cache, coreFlow),
-        InteractionEventHandler(kord, gateway, cache, coreFlow)
-    )
-
-    suspend fun start() = gateway.events
-        .buffer(Channel.UNLIMITED)
-        .onEach { (event, _, shard) -> dispatch(event, shard) }
-        .launchIn(kord)
-
-    private suspend fun dispatch(event: Event, shard: Int) {
-        runCatching {
-            listeners.forEach { it.handle(event, shard) }
-        }.onFailure {
-            logger.error(it)
-        }
-    }
-
-}
+}
```

Comment:
This is a really awkward interface for third parties to implement because it doesn't really hint at what it's supposed to do.

Assuming on GatewayEvent creates one CoreEvent, we can simplify the implementation to:
```kotlin
interface GatewayEventInterceptor {
    suspend fun handle(event: GatewayEvent, kord: Kord) : CoreEvent
}
```

⁰ cache and gateway can be retrieved from the kord instance
¹ This assumes that every gateway event produces exactly one core event, which may hold back flexibility

we can then move the launching logic over to Kord.

Reply:
This change definitly would make sense, it also keeps the flexibility.

Reply:
Though, having a custom "start" method gives the flexibility to read from a custom type of events from the Gateway, I think.